src/Dapper.AOT.Analyzers/Internal/Roslyn/TypeSymbolExtensions.cs

```diff
@@ -0,0 +1,171 @@
+﻿using Microsoft.CodeAnalysis;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Dapper.Internal.Roslyn;
+
+internal static class TypeSymbolExtensions
+{
+    public static bool InNamespace(this ITypeSymbol? typeSymbol, string @namespace)
+    {
+        if (typeSymbol is null) return false;
+        return typeSymbol.ContainingNamespace?.ToDisplayString() == @namespace;
+    }
+
+    public static string? GetTypeDisplayName(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return null;
+        if (typeSymbol.IsAnonymousType == true) return "object?";
+        return "global::" + typeSymbol.ToDisplayString();
+    }
+    public static string? GetContainingTypeDisplayName(this ITypeSymbol? typeSymbol, int typeArgIndex = 0)
+    {
+        var containingTypeSymbol = typeSymbol.GetContainingTypeSymbol(typeArgIndex);
+        return containingTypeSymbol.GetTypeDisplayName();
+    }
+    public static ITypeSymbol? GetContainingTypeSymbol(this ITypeSymbol? typeSymbol, int typeArgIndex = 0)
+    {
+        if (typeSymbol is null) return null;
+        if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+        {
+            var typeArguments = namedTypeSymbol.TypeArguments;
+            if (typeArgIndex >= typeArguments.Length) return null;
+            return typeArguments[typeArgIndex];
+        }
+        if (typeSymbol is IArrayTypeSymbol arrayTypeSymbol)
+        {
+            return arrayTypeSymbol.ElementType;
+        }
+
+        return null;
+    }
+
+    /// <returns>
+    /// True, if passed <param name="typeSymbol"/> represents <see cref="ImmutableArray{T}"/>.
+    /// False otherwise
+    /// </returns>
+    public static bool IsImmutableArray(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return false;
+        if (typeSymbol is not INamedTypeSymbol namedTypeSymbol) return false;
+
+        return namedTypeSymbol.TypeKind == TypeKind.Struct
+            && namedTypeSymbol.InNamespace("System.Collections.Immutable")
+            && namedTypeSymbol.Arity == 1
+            && namedTypeSymbol.Name == "ImmutableArray";
+    }
+
+    /// <returns>
+    /// True, if passed <param name="typeSymbol"/> represents array.
+    /// False otherwise
+    /// </returns>
+    public static bool IsArray(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return false;
```

Comment:
we don't want to check just for "is an array", as there exists two kinds of 1-dimensional array, even with zero bounds; I *think* we specifically want to test for `IsSZArray`,  aka "vector"; this actually makes the test a lot simpler - just:

return typeSymbol is IArrayTypeSymbol { IsSZArray : true }

(note: I don't think you can conveniently express the other array kind in C#, but: it absolutely exists)

Reply:
specifically, the difference is `int[]` vs `int[*]`, but you can't actually express `int[*]` directly in C#; despite that, I'd just rather do the more correct test, since an `int[*]` does not cast to `int[]`

Reply:
agree! done